handler/src/test/java/io/netty/handler/ssl/OpenSslEngineTest.java

```diff
@@ -1117,6 +1123,148 @@ public boolean permits(
         }
     }
 
+    @Test
+    public void testExtractMasterkeyWorksCorrectly() throws Exception {
+
+        SelfSignedCertificate cert = new SelfSignedCertificate();
+        SslContext serverContext = SslContextBuilder.forServer(cert.key(), cert.cert())
+                .sslProvider(SslProvider.OPENSSL).build();
+        final SSLEngine serverEngine =
+                serverContext.newEngine(UnpooledByteBufAllocator.DEFAULT);
+
+        SslContext clientContext = SslContextBuilder.forClient()
+                .trustManager(cert.certificate())
+                .sslProvider(SslProvider.OPENSSL).build();
+        final SSLEngine clientEngine =
+                clientContext.newEngine(UnpooledByteBufAllocator.DEFAULT);
+
+        //lets set the cipher suite to a specific one with DHE
+
+        //https://www.ietf.org/rfc/rfc5289.txt
+        //For cipher suites ending with _SHA256, the PRF is the TLS PRF
+        //[RFC5246] with SHA-256 as the hash function.  The MAC is HMAC
+        //[RFC2104] with SHA-256 as the hash function.
+        clientEngine.setEnabledCipherSuites(new String[]{ "TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA256"});
+        serverEngine.setEnabledCipherSuites(new String[]{ "TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA256"});
+
+        int appBufferMax = clientEngine.getSession().getApplicationBufferSize();
+        int netBufferMax = clientEngine.getSession().getPacketBufferSize();
+
+        /*
+         * We'll make the input buffers a bit bigger than the max needed
+         * size, so that unwrap()s following a successful data transfer
+         * won't generate BUFFER_OVERFLOWS.
+         */
+        ByteBuffer clientIn = ByteBuffer.allocate(appBufferMax + 50);
+        ByteBuffer serverIn = ByteBuffer.allocate(appBufferMax + 50);
+
+        ByteBuffer cTOs = ByteBuffer.allocate(netBufferMax);
+        ByteBuffer sTOc = ByteBuffer.allocate(netBufferMax);
+
+        ByteBuffer clientOut = ByteBuffer.wrap("Hi Server, I'm Client".getBytes());
+        ByteBuffer serverOut = ByteBuffer.wrap("Hello Client, I'm Server".getBytes());
+
+        // This implementation is largely imitated from
+        // https://docs.oracle.com/javase/8/docs/technotes/guides/security/jsse/samples/sslengine/SSLEngineSimpleDemo.java
+        // It has been simplified however without the need for running delegation tasks, as of now,
+        // OpenSslEngine does not do unwrap/wrap non-blocking
+
+        // Do handshake for SSL
+        // A typical handshake will usually contain the following steps:
+        // 1. wrap:     ClientHello
+        // 2. unwrap:   ServerHello/Cert/ServerHelloDone
+        // 3. wrap:     ClientKeyExchange
+        // 4. wrap:     ChangeCipherSpec
+        // 5. wrap:     Finished
+        // 6. unwrap:   ChangeCipherSpec
+        // 7. unwrap:   Finished
+
+        //set a for loop; instead of a while loop to guarantee we quit out eventually
+        boolean asserted = false;
+        for (int i = 0; i < 1000; i++) {
```

Comment:
why not use an endless loop and just configure a timeout ?

Reply:
Personal preference ? :)